### ISSUE TITLE
deno run: error when trying to import ESM module using require()

### ISSUE THREAD

**Issue by toxeus** [denoland/deno#17448]
When I run

```bash
deno run --node-modules-dir --allow-read --allow-env --allow-write npm:eslint index.ts
```

I get the following error

```text
Oops! Something went wrong! :(

ESLint: 8.32.0

Error: Trying to import ESM module: /home/runner/work/issue_repro/issue_repro/node_modules/.deno/espree@9.4.1/node_modules/espree/espree.js using require()
    at Object.Module._extensions..js (deno:ext/node/02_require.js:776:15)
    at Module.load (deno:ext/node/02_require.js:658:34)
    at Function.Module._load (deno:ext/node/02_require.js:515:14)
    at Module.require (deno:ext/node/02_require.js:680:21)
    at require (deno:ext/node/02_require.js:820:18)
    at Object.<anonymous> (file:///home/runner/work/issue_repro/issue_repro/node_modules/.deno/eslint@8.32.0/node_modules/eslint/lib/linter/linter.js:17:14)
    at Object.<anonymous> (file:///home/runner/work/issue_repro/issue_repro/node_modules/.deno/eslint@8.32.0/node_modules/eslint/lib/linter/linter.js:2084:4)
    at Module._compile (deno:ext/node/02_require.js:747:36)
    at Object.Module._extensions..js (deno:ext/node/02_require.js:780:12)
    at Module.load (deno:ext/node/02_require.js:658:34)
```

On top of that the command fails with a successful return code (`0`). So in the CI it shows as a success.

Here's a repro case: https://github.com/toxeus/issue_repro
And here is the the entire command output: https://github.com/toxeus/issue_repro/actions/runs/3932589655/jobs/6725334112

**Comment by bartlomieju** [comment]
Thanks for the report, seems like a bug in our `require()` implementation. I will take a stab at fixing it.

**Comment by jespertheend** [comment]
Fwiw, I'm successfully running eslint through Deno by wrapping their node api.
You can find an example here:
https://github.com/rendajs/Renda/blob/82b7f8bbbf662bf576c9de6f3af838c57c55e221/scripts/lint.js
The script contains a bunch of code you probably don't need, but it's a good proof of concept.

**Comment by toxeus** [comment]
Thanks for the feedback @jespertheend. The code path must be different for `deno run npm::eslint` then but should be easy to fix since there is a code path that already works. Might be a valuable insight for @bartlomieju :)

For deno to get traction the transition from node needs to be smooth. People will not want to exercise another migration in the style of Python2 -> Python3 voluntarily I fear.

**Comment by bartlomieju** [comment]
It appears we're not correctly using "conditional exports" in the require implementation and thus trying to load an ESM file instead of CJS one. Thanks for bringing this back up @toxeus, I'll fix it over the weekend.

**Comment by bartlomieju** [comment]
I debugged the problem and found the root cause, but the fix isn't there yet. I might have to revisit this after releasing 1.31 this week.

**Comment by kt3k** [comment]
closed by #17896
